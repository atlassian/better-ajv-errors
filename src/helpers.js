import {
  getChildren,
  getErrors,
  getSiblings,
  isAnyOfError,
  isEnumError,
  isRequiredError,
  concatAll,
  notUndefined,
} from './utils';
import {
  AdditionalPropValidationError,
  RequiredValidationError,
  EnumValidationError,
  DefaultValidationError,
} from './validation-errors';
import { debug } from './debug';

const JSON_POINTERS_REGEX = /\/[\w_-]+(\/\d+)?/g;

// Make a tree of errors from ajv errors array
export function makeTree(ajvErrors = []) {
  const root = { children: {} };
  ajvErrors.forEach(ajvError => {
    const { dataPath } = ajvError;

    // `dataPath === ''` is root
    const paths = dataPath === '' ? [''] : dataPath.match(JSON_POINTERS_REGEX);
    paths &&
      paths.reduce((obj, path, i) => {
        obj.children[path] = obj.children[path] || { children: {}, errors: [] };
        if (i === paths.length - 1) {
          obj.children[path].errors.push(ajvError);
        }
        return obj.children[path];
      }, root);
  });
  return root;
}

export function filterRedundantErrors(root, parent, key) {
  /**
   * If there is a `required` error then we can just skip everythig else.
   * And, also `required` should have more priority than `anyOf`. @see #8
   */
  getErrors(root).forEach(error => {
    if (isRequiredError(error)) {
      root.errors = [error];
      root.children = {};
    }
  });

  /**
   * If there is an `anyOf` error that means we have more meaningful errors
   * inside children. So we will just remove all errors from this level.
   */
  // TODO: Need to check children too. There might be no children :(
  if (getErrors(root).some(isAnyOfError)) {
    delete root.errors;
  }

  /**
   * If all errors are `enum` and siblings have any error then we can safely
   * ignore the node.
   *
   * **CAUTION**
   * Need explicit `root.errors` check because `[].every(fn) === true`
   * https://en.wikipedia.org/wiki/Vacuous_truth#Vacuous_truths_in_mathematics
   */
  if (root.errors && getErrors(root).every(isEnumError)) {
    if (
      getSiblings(parent)(root)
        // Remove any reference which becomes `undefined` later
        .filter(notUndefined)
        .some(getErrors)
    ) {
      delete parent.children[key];
    }
  }

  Object.entries(root.children).forEach(([key, child]) =>
    filterRedundantErrors(child, root, key)
  );
}

export function createErrorInstances(root, indent) {
  const errors = getErrors(root);
  if (errors.length && errors.every(isEnumError)) {
    const allowedValues = concatAll([])(
      errors.map(e => e.params.allowedValues)
    );
    const error = errors[0];
    return [
      new EnumValidationError(
        {
          ...error,
          params: { allowedValues },
        },
        indent
      ),
    ];
  } else {
    return concatAll(
      errors.reduce((ret, error) => {
        switch (error.keyword) {
          case 'additionalProperties':
            return ret.concat(new AdditionalPropValidationError(error, indent));
          case 'required':
            return ret.concat(new RequiredValidationError(error, indent));
          default:
            return ret.concat(new DefaultValidationError(error, indent));
        }
      }, [])
    )(
      getChildren(root).map(child => {
        return createErrorInstances(child, indent);
      })
    );
  }
}

export default (ajvErrors, indent) => {
  const tree = makeTree(ajvErrors || []);
  filterRedundantErrors(tree);
  const errors = createErrorInstances(tree, indent);
  if (!errors) {
    debug(JSON.stringify(tree));
  }
  return errors;
};
