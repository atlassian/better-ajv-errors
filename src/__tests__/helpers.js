import { makeTree, filterRedundantErrors } from '../helpers';

describe('makeTree', () => {
  it('works on empty array', async () => {
    expect(makeTree([])).toMatchInlineSnapshot(`
      Object {
        "children": Object {},
      }
    `);
  });

  it('works on root dataPath', async () => {
    expect(makeTree([{ dataPath: '' }])).toMatchInlineSnapshot(`
      Object {
        "children": Object {
          "": Object {
            "children": Object {},
            "errors": Array [
              Object {
                "dataPath": "",
              },
            ],
          },
        },
      }
    `);
  });

  it('works on nested dataPath', async () => {
    expect(makeTree([{ dataPath: '/root/child' }])).toMatchInlineSnapshot(`
      Object {
        "children": Object {
          "/root": Object {
            "children": Object {
              "/child": Object {
                "children": Object {},
                "errors": Array [
                  Object {
                    "dataPath": "/root/child",
                  },
                ],
              },
            },
            "errors": Array [],
          },
        },
      }
    `);
  });

  it('works on array dataPath', async () => {
    expect(
      makeTree([{ dataPath: '/root/child/0' }, { dataPath: '/root/child/1' }])
    ).toMatchInlineSnapshot(`
      Object {
        "children": Object {
          "/root": Object {
            "children": Object {
              "/child/0": Object {
                "children": Object {},
                "errors": Array [
                  Object {
                    "dataPath": "/root/child/0",
                  },
                ],
              },
              "/child/1": Object {
                "children": Object {},
                "errors": Array [
                  Object {
                    "dataPath": "/root/child/1",
                  },
                ],
              },
            },
            "errors": Array [],
          },
        },
      }
    `);
  });

  it('works on array item dataPath', async () => {
    expect(
      makeTree([
        { dataPath: '/root/child/0/grand-child' },
        { dataPath: '/root/child/1/grand-child' },
      ])
    ).toMatchInlineSnapshot(`
      Object {
        "children": Object {
          "/root": Object {
            "children": Object {
              "/child/0": Object {
                "children": Object {
                  "/grand-child": Object {
                    "children": Object {},
                    "errors": Array [
                      Object {
                        "dataPath": "/root/child/0/grand-child",
                      },
                    ],
                  },
                },
                "errors": Array [],
              },
              "/child/1": Object {
                "children": Object {
                  "/grand-child": Object {
                    "children": Object {},
                    "errors": Array [
                      Object {
                        "dataPath": "/root/child/1/grand-child",
                      },
                    ],
                  },
                },
                "errors": Array [],
              },
            },
            "errors": Array [],
          },
        },
      }
    `);
  });
});

describe('filterRedundantErrors', () => {
  it('should prioritize required', async () => {
    const tree = {
      children: {
        a: {
          children: {
            b: {
              children: {},
              errors: [
                {
                  keyword: 'required',
                },
              ],
            },
          },
          errors: [
            {
              keyword: 'required',
            },
            {
              keyword: 'anyOf',
            },
            {
              keyword: 'enum',
            },
          ],
        },
      },
    };
    filterRedundantErrors(tree);
    expect(tree).toMatchInlineSnapshot(`
      Object {
        "children": Object {
          "a": Object {
            "children": Object {},
            "errors": Array [
              Object {
                "keyword": "required",
              },
            ],
          },
        },
      }
    `);
  });

  it('should handle anyOf', async () => {
    const tree = {
      children: {
        a: {
          children: {
            b: {
              children: {},
              errors: [
                {
                  keyword: 'required',
                },
              ],
            },
          },
          errors: [
            {
              keyword: 'anyOf',
            },
            {
              keyword: 'enum',
            },
          ],
        },
      },
    };
    filterRedundantErrors(tree);
    expect(tree).toMatchInlineSnapshot(`
      Object {
        "children": Object {
          "a": Object {
            "children": Object {
              "b": Object {
                "children": Object {},
                "errors": Array [
                  Object {
                    "keyword": "required",
                  },
                ],
              },
            },
          },
        },
      }
    `);
  });

  it('should handle enum', async () => {
    const tree = {
      children: {
        a: {
          children: {
            b: {
              children: {},
              errors: [
                {
                  keyword: 'enum',
                },
                {
                  keyword: 'enum',
                },
              ],
            },
          },
          errors: [
            {
              keyword: 'anyOf',
            },
            {
              keyword: 'additionalProperty',
            },
          ],
        },
      },
    };
    filterRedundantErrors(tree);
    expect(tree).toMatchInlineSnapshot(`
      Object {
        "children": Object {
          "a": Object {
            "children": Object {
              "b": Object {
                "children": Object {},
                "errors": Array [
                  Object {
                    "keyword": "enum",
                  },
                  Object {
                    "keyword": "enum",
                  },
                ],
              },
            },
          },
        },
      }
    `);
  });

  it('should handle enum - sibling', async () => {
    const tree = {
      children: {
        a1: {
          children: {},
          errors: [
            {
              keyword: 'enum',
            },
            {
              keyword: 'enum',
            },
          ],
        },
        a2: {
          children: {},
          errors: [
            {
              keyword: 'additionalProperty',
            },
          ],
        },
      },
    };
    filterRedundantErrors(tree);
    expect(tree).toMatchInlineSnapshot(`
      Object {
        "children": Object {
          "a2": Object {
            "children": Object {},
            "errors": Array [
              Object {
                "keyword": "additionalProperty",
              },
            ],
          },
        },
      }
    `);
  });

  it('should handle enum - sibling with nested error', async () => {
    const tree = {
      children: {
        a1: {
          children: {
            b1: {
              children: {},
              errors: [
                {
                  keyword: 'additionalProperty',
                },
              ],
            },
          },
          errors: [],
        },
        a2: {
          children: {},
          errors: [
            {
              keyword: 'enum',
            },
            {
              keyword: 'enum',
            },
          ],
        },
      },
    };
    filterRedundantErrors(tree);
    expect(tree).toMatchInlineSnapshot(`
      Object {
        "children": Object {
          "a1": Object {
            "children": Object {
              "b1": Object {
                "children": Object {},
                "errors": Array [
                  Object {
                    "keyword": "additionalProperty",
                  },
                ],
              },
            },
            "errors": Array [],
          },
        },
      }
    `);
  });

  it('should not remove anyOf errors if there are no children', async () => {
    const tree = {
      children: {
        '/object': {
          children: {
            '/type': {
              children: {},
              errors: [
                {
                  keyword: 'type',
                },
                {
                  keyword: 'type',
                },
                {
                  keyword: 'anyOf',
                },
              ],
            },
          },
          errors: [],
        },
      },
    };

    filterRedundantErrors(tree);
    expect(tree).toMatchInlineSnapshot(`
      Object {
        "children": Object {
          "/object": Object {
            "children": Object {
              "/type": Object {
                "children": Object {},
                "errors": Array [
                  Object {
                    "keyword": "type",
                  },
                  Object {
                    "keyword": "type",
                  },
                  Object {
                    "keyword": "anyOf",
                  },
                ],
              },
            },
            "errors": Array [],
          },
        },
      }
    `);
  });
});
