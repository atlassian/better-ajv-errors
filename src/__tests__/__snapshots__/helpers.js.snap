// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`filterRedundantErrors should handle anyOf 1`] = `
Object {
  "children": Object {
    "a": Object {
      "children": Object {
        "b": Object {
          "children": Object {},
          "errors": Array [
            Object {
              "keyword": "required",
            },
          ],
        },
      },
    },
  },
}
`;

exports[`filterRedundantErrors should handle enum - sibling 1`] = `
Object {
  "children": Object {
    "a2": Object {
      "children": Object {},
      "errors": Array [
        Object {
          "keyword": "additionalProperty",
        },
      ],
    },
  },
}
`;

exports[`filterRedundantErrors should handle enum 1`] = `
Object {
  "children": Object {
    "a": Object {
      "children": Object {
        "b": Object {
          "children": Object {},
          "errors": Array [
            Object {
              "keyword": "enum",
            },
            Object {
              "keyword": "enum",
            },
          ],
        },
      },
    },
  },
}
`;

exports[`filterRedundantErrors should prioritize required 1`] = `
Object {
  "children": Object {
    "a": Object {
      "children": Object {},
      "errors": Array [
        Object {
          "keyword": "required",
        },
      ],
    },
  },
}
`;

exports[`makeTree works on array dataPath 1`] = `
Object {
  "children": Object {
    "/root": Object {
      "children": Object {
        "/child/0": Object {
          "children": Object {},
          "errors": Array [
            Object {
              "dataPath": "/root/child/0",
            },
          ],
        },
        "/child/1": Object {
          "children": Object {},
          "errors": Array [
            Object {
              "dataPath": "/root/child/1",
            },
          ],
        },
      },
      "errors": Array [],
    },
  },
}
`;

exports[`makeTree works on array item dataPath 1`] = `
Object {
  "children": Object {
    "/root": Object {
      "children": Object {
        "/child/0": Object {
          "children": Object {
            "/grand-child": Object {
              "children": Object {},
              "errors": Array [
                Object {
                  "dataPath": "/root/child/0/grand-child",
                },
              ],
            },
          },
          "errors": Array [],
        },
        "/child/1": Object {
          "children": Object {
            "/grand-child": Object {
              "children": Object {},
              "errors": Array [
                Object {
                  "dataPath": "/root/child/1/grand-child",
                },
              ],
            },
          },
          "errors": Array [],
        },
      },
      "errors": Array [],
    },
  },
}
`;

exports[`makeTree works on empty array 1`] = `
Object {
  "children": Object {},
}
`;

exports[`makeTree works on nested dataPath 1`] = `
Object {
  "children": Object {
    "/root": Object {
      "children": Object {
        "/child": Object {
          "children": Object {},
          "errors": Array [
            Object {
              "dataPath": "/root/child",
            },
          ],
        },
      },
      "errors": Array [],
    },
  },
}
`;

exports[`makeTree works on root dataPath 1`] = `
Object {
  "children": Object {
    "": Object {
      "children": Object {},
      "errors": Array [
        Object {
          "dataPath": "",
        },
      ],
    },
  },
}
`;
